### PR TITLE
cherry-pick to main : fix setvar in the active transaction

### PR DESCRIPTION
**Pull request by daviszhen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #https://github.com/matrixorigin/MO-Cloud/issues/2343
issue #https://github.com/matrixorigin/MO-Cloud/issues/2048

## What this PR does / why we need it:

之前的设计，限制set 用在活跃事务中。原因是早期mo的bug奇多，为了避免踩坑，做了限制。
现在由于与mysql 应用的兼容性原因。尝试放开。

1，set @autocommit = ON:
从  OFF 到 ON，mo会commit事务。如果commit失败，autocommit 标志不会设置成功。

### CODE REVIEW

**Comment by matrix-meow** [comment]
@daviszhen Thanks for your contributions!

**Here are review comments for file pkg/frontend/mysql_cmd_executor.go:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
The body of the pull request provides some context about the changes made. It mentions that the previous design limited the use of "set" in active transactions due to bugs in the early version of the software. However, for compatibility with MySQL applications, the limitation is being removed. It also mentions a specific scenario where setting "@autocommit = ON" would cause the transaction to be committed and if the commit fails, the autocommit flag won't be set successfully.

Changes in pkg/frontend/mysql_cmd_executor.go:

1. In the `doSetVar` function, a new variable `oldValueRaw` is declared to store the old value of the session variable if the name is "autocommit". This is done to handle the scenario mentioned in the pull request body.

2. The `ses.SetAutocommit` function is modified to accept both the old and new values of the autocommit flag. This change is made to handle the scenario mentioned in the pull request body.

3. The check for `IsParameterModificationStatement(stmt)` is removed in the `canExecuteStatementInUncommittedTransaction` function. This check was previously used to prevent parameter modification statements in transactions, but it has been removed without any explanation in the pull request. This change should be reviewed and confirmed if it is intentional.

4. In the `executeStmt` function, the call to `ses.GetTxnHandler().disableStartStmt()` is removed. It is unclear why this change was made and if it is intentional. This change should be reviewed and confirmed if it is necessary.

Suggestions for improvement:

1. The pull request body should provide more detailed explanations for the changes made. It should explain why the limitation on using "set" in active transactions is being removed and how it improves compatibility with MySQL applications.

2. The removal of the check for `IsParameterModificationStatement(stmt)` in the `canExecuteStatementInUncommittedTransaction` function should be reviewed and confirmed if it is intentional. If it is not intentional, the check should be restored or an explanation should be provided for why it was removed.

3. The removal of the call to `ses.GetTxnHandler().disableStartStmt()` in the `executeStmt` function should be reviewed and confirmed if it is necessary. If it is not necessary, an explanation should be provided for why it was removed.

4. It would be helpful to include test cases for the changes made in the pull request to ensure that they are functioning as expected and to prevent regressions.

5. The code changes could benefit from more descriptive variable and function names to improve readability and maintainability.

Security issues:

No security issues were identified in the provided pull request.

**Here are review comments for file pkg/frontend/session_test.go:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
- The PR fixes an issue related to setting variables in an active transaction.
- The PR is marked as a bug fix.
- The PR references two issues: #2343 and #2048.
- The PR aims to remove the restriction on using the `set` command in an active transaction for compatibility with MySQL applications.
- The PR includes changes to the `pkg/frontend/session_test.go` file.

Changes in `pkg/frontend/session_test.go`:
- Line 259: The test case previously attempted to set autocommit to false and expected an error. This has been changed to set autocommit to true and expect no error.
- Line 269: The test case previously attempted to set autocommit to true and expected no error. This has been changed to set autocommit to false and expect no error.

Potential problems and suggestions:
1. In the body of the PR, the references to the issues are incorrect. The correct format for referencing issues is `#issue_number`, without the `https://github.com/` part. The correct references should be `#2343` and `#2048`. Suggestion: Update the references to the correct format.
2. The body of the PR is written in a language other than English, which may make it difficult for non-native speakers to understand. Suggestion: Translate the body of the PR into English for better clarity and comprehension.
3. The changes made in the `pkg/frontend/session_test.go` file seem to contradict the original intention of the test case. The original test case was checking for an error when setting autocommit to false, but the changes now set autocommit to true and expect no error. This change may introduce a regression or unintended behavior. Suggestion: Review the changes and ensure they align with the original intention of the test case.
4. The changes made in the `pkg/frontend/session_test.go` file could be optimized by removing unnecessary lines. For example, lines 269 and 269 are now identical, so one of them can be removed. Suggestion: Remove the duplicate line to improve code readability and maintainability.

Security issues:
No security issues were identified in the provided pull request.

**Here are review comments for file pkg/frontend/stmt_kind.go:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
The body of the pull request provides some context about the changes being made. It mentions that the previous design limited the use of the "set" command in active transactions due to bugs in the early version of the software. However, for compatibility reasons with MySQL applications, the limitation is being removed. It also mentions that setting "@autocommit = ON" will now commit the transaction, and if the commit fails, the autocommit flag will not be set successfully.

Changes in pkg/frontend/stmt_kind.go:
The changes in this file add support for executing the "set" command in an active transaction. The "statementCanBeExecutedInUncommittedTransaction" function now returns true for the "SetVar" statement, allowing it to be executed in an active transaction.

Potential Problems and Suggestions:

1. Lack of English translation in the body:
The body of the pull request is written in a language other than English, which may make it difficult for non-native speakers to understand. It is important to provide translations or explanations in English to ensure clarity and understanding for all contributors and reviewers.

2. Inconsistent use of issue links:
The body of the pull request includes issue links in a format that is not recognized by GitHub. Instead of using "issue #https://github.com/matrixorigin/MO-Cloud/issues/2343", it should be formatted as "Fixes #2343" or "Closes #2343" to automatically close the linked issue when the pull request is merged.

3. Lack of unit tests:
The changes made in the pull request do not include any unit tests to verify the functionality of the "SetVar" statement in an active transaction. It is important to add appropriate unit tests to ensure that the changes work as intended and to prevent regressions in the future.

4. Security concern with setting "@autocommit = ON":
The pull request mentions that setting "@autocommit = ON" will now commit the transaction. This can be a security concern as it allows the transaction to be automatically committed without explicit user action. It is important to carefully consider the implications of this change and ensure that it does not introduce any vulnerabilities or unintended behavior.

Optimization Suggestions:

1. Improve code readability:
The code changes in pkg/frontend/stmt_kind.go could be improved for better readability. Consider adding comments to explain the purpose and behavior of the code. Additionally, consider using more descriptive variable and function names to make the code easier to understand.

2. Consolidate related changes:
If there are other changes related to the "SetVar" statement or active transactions, it would be beneficial to include them in the same pull request. This helps to keep related changes together and makes it easier for reviewers to understand the overall impact of the changes.

Overall, the pull request addresses a bug and adds support for executing the "SetVar" statement in an active transaction. However, there are some areas that need improvement, such as providing English translations, adding unit tests, addressing security concerns, and optimizing the code changes for readability.

**Here are review comments for file pkg/frontend/txn.go:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
- The PR is a bug fix.
- It fixes two issues: #2343 and #2048.
- The PR aims to address a limitation in the previous design where the 'set' command was restricted in active transactions. This limitation was implemented to avoid potential bugs in the early stages of the project. However, for compatibility reasons with MySQL applications, the restriction is being lifted.
- The specific change in this PR is related to the `SetAutocommit` function in the `txn.go` file.

Changes in `pkg/frontend/txn.go`:
- The `SetAutocommit` function has been modified.
- The function now takes two boolean parameters: `old` and `on`.
- The function now includes additional logic to handle different scenarios when setting the autocommit value.
- If the `old` value is `false` and the `on` value is `true`, it means that autocommit is being activated. In this case, the function attempts to commit the active transaction. If the commit fails, the `OPTION_AUTOCOMMIT` flag is cleared, and the error is returned. If the commit succeeds, the `OPTION_BEGIN` and `OPTION_NOT_AUTOCOMMIT` flags are cleared, and the `SERVER_STATUS_AUTOCOMMIT` flag is set.
- If the `old` value is `true` and the `on` value is `false`, it means that autocommit is being deactivated. In this case, the `SERVER_STATUS_AUTOCOMMIT` flag is cleared, the `OPTION_NOT_AUTOCOMMIT` flag is set, and no further action is taken.

Potential problems and suggestions:
1. In the PR body, the links to the issues are incorrect. They should be in the format `#2343` and `#2048` without the `https://github.com/` part.
2. The PR body is written in a language other than English, which may make it difficult for non-native speakers to understand. It would be better to provide an English translation or use English in the first place.
3. The `SetAutocommit` function has been modified to handle different scenarios, but the code could benefit from more comments explaining the logic and the purpose of each step. This would make it easier for future developers to understand and maintain the code.
4. The function signature has changed, adding a new parameter `old`. It would be helpful to update the function documentation to explain the purpose and meaning of this parameter.
5. The error handling in the `SetAutocommit` function could be improved. Currently, if the commit fails, the error is returned, but there is no indication of what caused the commit to fail. It would be helpful to provide a more specific error message or handle the error in a way that provides more context.
6. It's unclear why the `ses.ClearOptionBits(OPTION_AUTOCOMMIT)` is called when the commit fails. This action seems redundant since the `OPTION_AUTOCOMMIT` flag is already cleared in the `if err != nil` block. It should be verified if this line is necessary or can be removed.
7. The code could benefit from additional testing to ensure that all scenarios are properly handled and that the behavior matches the intended logic.

Optimization suggestions:
1. The function could be further optimized by removing the unnecessary `else` statement. Since the `if` condition is already checking for the opposite scenario, there is no need for an `else` block. This would simplify the code and make it more readable.
2. Consider using more descriptive variable names to improve code readability. For example, instead of `old` and `on`, consider using names like `previousAutocommit` and `newAutocommit` to make the code more self-explanatory.

Overall, the changes in this pull request address a bug and aim to improve compatibility with MySQL applications. However, there are some potential problems and areas for improvement that should be addressed before merging the changes.

**Here are review comments for file test/distributed/cases/optimistic/autocommit.result:**
Title: cherry-pick to main : fix setvar in the active transaction

Body:
- This pull request fixes an issue related to the `setvar` command in an active transaction.
- The previous design restricted the use of `setvar` in an active transaction due to numerous bugs in the early version of the software. This restriction was put in place to avoid potential issues.
- However, for compatibility reasons with MySQL applications, the restriction is being lifted.

Changes in the file `test/distributed/cases/optimistic/autocommit.result`:
- Line 29: The table name `tables_in_autocommit` is changed to `Tables_in_autocommit`.
- Line 46: The error message "internal error: Uncommitted transaction exists. Please commit or rollback first." is removed.
- Line 63: The table name `tables_in_autocommit` is changed to `Tables_in_autocommit`.

Overall, this pull request aims to fix the issue related to the `setvar` command in an active transaction and make the software more compatible with MySQL applications.

Potential Problems and Suggestions:
1. Security Issue: The pull request does not mention any security-related changes. It is important to ensure that the changes made in this pull request do not introduce any security vulnerabilities. It is recommended to conduct a thorough security review of the code changes to identify and address any potential security issues.

2. Lack of Detailed Explanation: The pull request description lacks detailed explanation about the specific bug being fixed and the reason for lifting the restriction on the `setvar` command. It would be helpful to provide more information about the bug and the impact it had on the software. Additionally, explaining the compatibility issues with MySQL applications and the decision to lift the restriction would provide better context for the reviewers.

3. Inconsistent Table Name: The table name `tables_in_autocommit` is changed to `Tables_in_autocommit` in lines 29 and 63 of the file `test/distributed/cases/optimistic/autocommit.result`. It is important to ensure consistent naming conventions throughout the codebase. It is recommended to choose a consistent naming convention for table names and apply it consistently across the codebase.

Optimization Suggestions:
1. Code Refactoring: The pull request does not mention any code refactoring changes. However, it is always beneficial to review the code for potential refactoring opportunities. This can include improving code readability, removing duplicate code, optimizing performance, and following best practices. It is recommended to conduct a code review to identify any areas of improvement and consider refactoring the code if necessary.

2. Test Coverage: The pull request does not mention any changes related to test coverage. It is important to ensure that the changes made in this pull request are adequately tested to prevent regressions and ensure the overall stability of the software. It is recommended to review the existing test suite and consider adding additional tests if necessary to cover the changes made in this pull request.

Overall, this pull request addresses a specific bug related to the `setvar` command in an active transaction and aims to improve compatibility with MySQL applications. However, it is important to address the potential security issue, provide more detailed explanations, ensure consistent naming conventions, and consider code refactoring and test coverage improvements.

**Here are review comments for file test/distributed/cases/optimistic/autocommit_1.result:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
The body of the pull request provides some context and explains the motivation behind the changes. It mentions that the previous design limited the use of the "set" command in active transactions due to bugs in the early version of the software. However, for compatibility with MySQL applications, the limitation is being removed. 

Changes:
The changes in the pull request are made to the file "test/distributed/cases/optimistic/autocommit_1.result". The changes include modifying the expected results of certain queries related to the "autocommit" setting. The modified queries no longer return an "internal error" message when executed within an active transaction.

Potential Problems and Suggestions:

1. Inconsistent language: The body of the pull request is written in a mix of English and Chinese. It would be better to use a consistent language throughout the pull request to ensure clarity and readability for all contributors.

2. Lack of detailed explanation: The body of the pull request briefly mentions the motivation behind the changes but does not provide a detailed explanation of the problem being addressed. It would be helpful to include more information about the specific bugs or issues that were encountered and how the changes fix them.

3. Incomplete issue references: The pull request mentions two issues that it fixes, but the issue references provided are not in the correct format. Instead of using "issue #" followed by a URL, the references should be in the format "#issue_number". For example, "#2343" instead of "https://github.com/matrixorigin/MO-Cloud/issues/2343". This will ensure that the issues are properly linked and tracked.

4. Lack of test cases: The pull request only includes changes to the expected results of existing test cases. It would be beneficial to also include new test cases that specifically cover the scenarios being addressed by the changes. This will help ensure that the changes are thoroughly tested and that any potential regressions are caught.

5. Security implications: Based on the provided information, there are no apparent security issues introduced by the changes in this pull request.

Optimization Suggestions:

1. Clarify commit failure behavior: The pull request mentions that if the commit fails when setting "@autocommit = ON", the "autocommit" flag will not be set successfully. It would be helpful to clarify what happens in this case. Does the transaction remain active or is it automatically rolled back? Providing this information in the comments or documentation will help users understand the expected behavior.

2. Improve error messages: The modified queries no longer return an "internal error" message when executed within an active transaction. However, it would be beneficial to provide more informative error messages that clearly indicate the reason for the error. This will help users troubleshoot and understand the issue more easily.

3. Consider adding rollback statements: Since the changes allow the use of "set" commands in active transactions, it may be necessary to add rollback statements in the test cases to ensure that the database state is properly reset after each test. This will help avoid potential side effects and ensure the reliability of the test suite.

Overall, the pull request addresses a limitation in the software and improves compatibility with MySQL applications. However, there are some areas that need improvement, such as providing more detailed explanations, using consistent language, and adding new test cases. Additionally, clarifying the behavior in case of commit failure and improving error messages will enhance the usability of the software.

**Here are review comments for file test/distributed/cases/optimistic/autocommit_1.sql:**
Review:

Title: cherry-pick to main : fix setvar in the active transaction

Body:
- The PR is labeled as a bug fix.
- It references two issues that this PR fixes.
- The PR aims to address a limitation in the previous design where the "set" command was restricted to active transactions. The restriction was put in place to avoid potential bugs, but now it is being relaxed for compatibility with MySQL applications.
- The PR includes changes to the file `test/distributed/cases/optimistic/autocommit_1.sql`.

Changes:
- The changes in the file `test/distributed/cases/optimistic/autocommit_1.sql` involve modifying the comments and queries related to the `set @@autocommit` command.
- The comments have been updated to reflect the expected behavior of the `set @@autocommit` command.
- The queries have been modified to check the value of `@@autocommit` after the `set` command and verify that there are no errors.

Potential Problems and Suggestions:
1. Inconsistent language: The comments in the file `test/distributed/cases/optimistic/autocommit_1.sql` use a mix of English and Chinese characters. It is recommended to use a consistent language (preferably English) throughout the codebase to ensure readability and maintainability.

2. Incomplete error handling: The comments indicate that there should be no errors when executing the `set @@autocommit` command. However, there is no explicit error handling or validation in the code. It is recommended to add appropriate error handling to handle any potential errors that may occur during the execution of the `set` command.

3. Lack of test cases: The changes in the file `test/distributed/cases/optimistic/autocommit_1.sql` seem to be focused on verifying the behavior of the `set @@autocommit` command. However, there are no additional test cases added to cover different scenarios or edge cases. It is recommended to add more comprehensive test cases to ensure the correctness and robustness of the code.

4. Security concern: The PR does not mention any security-related changes, but it is important to consider potential security implications when modifying code related to transactions and autocommit behavior. It is recommended to conduct a thorough security review to ensure that the changes do not introduce any vulnerabilities or security risks.

Optimization Suggestions:
1. Code organization: The changes in the file `test/distributed/cases/optimistic/autocommit_1.sql` could be organized in a more structured and readable manner. Consider grouping the queries and comments related to each test case together to improve clarity and maintainability.

2. Consistent formatting: Ensure consistent formatting of the code, including indentation, spacing, and line breaks. Consistent formatting improves readability and makes it easier to understand and review the code.

3. Use descriptive comments: The comments in the file `test/distributed/cases/optimistic/autocommit_1.sql` could be more descriptive and provide additional context about the purpose and expected behavior of each test case. This would make it easier for future developers to understand the code and its intent.

4. Consider using a testing framework: Instead of manually writing and executing queries, consider using a testing framework that provides more structured and automated testing capabilities. This would make it easier to write and maintain test cases, and also provide better reporting and error handling.

Overall, the changes in the pull request seem to address a specific limitation and aim to improve compatibility with MySQL applications. However, there are some areas that need attention, such as inconsistent language, incomplete error handling, lack of test cases, and potential security concerns. Additionally, optimizing the code organization and formatting can improve readability and maintainability.

**Here are review comments for file test/distributed/cases/optimistic/transaction_enhance.result:**
Title: cherry-pick to main : fix setvar in the active transaction

Body:
The pull request fixes an issue related to the `setvar` command in an active transaction. The previous design restricted the use of `setvar` in active transactions due to numerous bugs in the early version of the software. However, for compatibility reasons with MySQL applications, the restriction is being lifted. 

Changes Made:
The changes made in the pull request are in the file `test/distributed/cases/optimistic/transaction_enhance.result`. The changes include adding empty lines and modifying the indentation of the code. There are also changes made to the output of some SQL queries.

Review:

1. The title of the pull request is not very descriptive. It would be helpful to include more information about the specific issue being fixed in the title.

2. The body of the pull request is written in a language that is not clear. It would be beneficial to provide a clear explanation of the issue being fixed and the reason behind the fix. Additionally, it would be helpful to provide more context about the compatibility issues with MySQL applications.

3. The changes made in the file `test/distributed/cases/optimistic/transaction_enhance.result` seem to be unrelated to the issue described in the body of the pull request. It is unclear why these changes were made and how they are related to the fix for the `setvar` command.

4. It is important to ensure that the changes made in the pull request do not introduce any security vulnerabilities. It would be beneficial to review the changes for any potential security issues and address them if necessary.

Suggestions:

1. Update the title of the pull request to provide a more descriptive summary of the issue being fixed.

2. Rewrite the body of the pull request to provide a clear explanation of the issue being fixed, the reason behind the fix, and the compatibility issues with MySQL applications.

3. Provide more context about the changes made in the file `test/distributed/cases/optimistic/transaction_enhance.result` and explain how they are related to the fix for the `setvar` command.

4. Review the changes for any potential security issues and address them if necessary.

5. Consider optimizing the changes made in the pull request by removing any unrelated or unnecessary changes. This will help to keep the codebase clean and maintainable.

**Here are review comments for file test/distributed/cases/pessimistic_transaction/isolation_2.result:**
Title: cherry-pick to main : fix setvar in the active transaction

Body:
The pull request fixes an issue related to the `setvar` command in an active transaction. Previously, the `setvar` command was restricted to be used only in active transactions due to numerous bugs in the early version of MO. However, for compatibility reasons with MySQL applications, the restriction is being lifted. The pull request aims to address this issue.

Changes in the file `test/distributed/cases/pessimistic_transaction/isolation_2.result`:
- Line 343: The semicolon (;) after the `delete` statement has been removed.

Review:

1. The title of the pull request is clear and concise, indicating that it fixes an issue related to the `setvar` command in an active transaction.

2. The body of the pull request provides some context about the previous design decision and the reason for lifting the restriction on the `setvar` command. However, the body is written in a language other than English, which may make it difficult for non-native speakers to understand. It would be helpful to provide an English translation or summary of the body.

3. The changes in the file `test/distributed/cases/pessimistic_transaction/isolation_2.result` seem to be unrelated to the issue mentioned in the pull request. The only change is the removal of a semicolon (;) after the `delete` statement. It is unclear why this change is necessary or how it relates to the fix for the `setvar` command.

Suggestions:

1. Provide an English translation or summary of the body of the pull request to ensure that all reviewers can understand the context and purpose of the changes.

2. Clarify the relationship between the changes in the file `test/distributed/cases/pessimistic_transaction/isolation_2.result` and the fix for the `setvar` command. If the changes are unrelated, consider creating a separate pull request for them.

3. Optimize the changes made in the pull request by ensuring that they are focused on addressing the specific issue with the `setvar` command. Remove any unrelated changes to improve the clarity and maintainability of the codebase.

**Here are review comments for file test/distributed/cases/statement_query_type/statement_query_type.result:**
Title: cherry-pick to main : fix setvar in the active transaction

Body:
- This pull request fixes a bug related to the "setvar" command in an active transaction.
- The previous design limited the use of "setvar" in active transactions due to numerous bugs in the early version of MO. This limitation was put in place to avoid potential issues.
- However, for compatibility reasons with MySQL applications, this pull request attempts to remove the limitation.
- Specifically, when setting "@autocommit" to "ON", MO will commit the transaction. If the commit fails, the "autocommit" flag will not be set successfully.

Changes in the file "test/distributed/cases/statement_query_type/statement_query_type.result":
- Line 202: The test case previously expected an error message "internal error: Uncommitted transaction exists. Please commit or rollback first." when executing the "set @a=2;" command in an active transaction.
- The change in this pull request removes the expected error message, indicating that the fix allows the "set @a=2;" command to be executed successfully in an active transaction.

Review:
- The title of the pull request is clear and concise, indicating that it fixes an issue related to the "setvar" command in an active transaction.
- The body of the pull request provides some context and explains the motivation behind the changes.
- The changes in the file "test/distributed/cases/statement_query_type/statement_query_type.result" appear to be correct and aligned with the fix described in the pull request.

Suggestions:
1. Security: It's important to ensure that the removal of the limitation on the "setvar" command in active transactions does not introduce any security vulnerabilities. It would be beneficial to conduct a thorough security review to identify any potential risks and address them accordingly.

2. Code Optimization: The changes in the pull request seem to be focused on fixing a specific bug. However, it would be helpful to review the code and identify any other potential improvements or optimizations that can be made while addressing this issue. This could include refactoring code, improving performance, or enhancing error handling.

3. Testing: It would be valuable to add additional test cases to cover different scenarios related to the "setvar" command in active transactions. This will help ensure that the fix is robust and does not introduce regressions.

4. Documentation: It would be beneficial to update the documentation to reflect the changes made in this pull request. This will help other developers understand the behavior of the "setvar" command in active transactions and any limitations or considerations they need to be aware of.

Overall, the pull request seems to address the reported bug effectively. However, it is important to thoroughly review the code and consider the suggestions provided to ensure the fix is comprehensive and optimized.

**Review by qingxinhome — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**
